# NASA Exoplanet Data Cleaning Pipeline
## Datasets: Kepler, K2, and TESS

This notebook will explore, clean, and potentially combine three NASA exoplanet discovery datasets:
- **Kepler**: Original Kepler mission data
- **K2**: Extended K2 mission data
- **TESS**: Transiting Exoplanet Survey Satellite data

### Analysis Plan:
1. **Setup & Import** - Load required libraries
2. **Data Loading** - Load all three CSV files
3. **Individual Exploration** - Examine each dataset's structure, statistics, and quality
4. **Schema Comparison** - Identify common/unique columns across datasets
5. **Data Cleaning** - Handle missing values, duplicates, and inconsistencies
6. **Combination Strategy** - Determine if/how to combine datasets
7. **Export** - Save cleaned data for analysis

## 1. Setup & Imports

In [5]:
# Import required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from pathlib import Path
import warnings

# Configure display settings
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)
pd.set_option('display.float_format', lambda x: '%.3f' % x)
warnings.filterwarnings('ignore')

# Set visualization style
sns.set_style('whitegrid')
plt.rcParams['figure.figsize'] = (12, 6)

print("✓ Libraries imported successfully")

✓ Libraries imported successfully


## 2. Load Datasets

Load all three NASA exoplanet datasets from the raw data folder.

In [8]:
# Define data paths
data_dir = Path('../data/raw')

# Load datasets
# NASA Exoplanet Archive files use '#' for comment lines
print("Loading datasets...")
kepler_df = pd.read_csv(data_dir / 'kepler.csv', comment='#', on_bad_lines='skip')
k2_df = pd.read_csv(data_dir / 'k2.csv', comment='#', on_bad_lines='skip')
tess_df = pd.read_csv(data_dir / 'tess.csv', comment='#', on_bad_lines='skip')

print(f"✓ Kepler dataset loaded: {kepler_df.shape[0]:,} rows × {kepler_df.shape[1]} columns")
print(f"✓ K2 dataset loaded: {k2_df.shape[0]:,} rows × {k2_df.shape[1]} columns")
print(f"✓ TESS dataset loaded: {tess_df.shape[0]:,} rows × {tess_df.shape[1]} columns")

Loading datasets...
✓ Kepler dataset loaded: 9,564 rows × 141 columns
✓ K2 dataset loaded: 4,004 rows × 295 columns
✓ TESS dataset loaded: 7,703 rows × 87 columns
✓ Kepler dataset loaded: 9,564 rows × 141 columns
✓ K2 dataset loaded: 4,004 rows × 295 columns
✓ TESS dataset loaded: 7,703 rows × 87 columns


## 3. Explore Individual Datasets

### 3.1 Kepler Dataset Exploration

In [14]:
# Kepler: Basic information
print("=" * 80)
print("KEPLER DATASET EXPLORATION")
print("=" * 80)

print(f"\nDataset shape: {kepler_df.shape[0]:,} rows × {kepler_df.shape[1]} columns\n")

print("Column names and types:")
print(kepler_df.dtypes)

print("\n" + "-" * 80)
print("First few rows:")
display(kepler_df.head())

print("\n" + "-" * 80)
print("Missing values:")
missing_kepler = kepler_df.isnull().sum()
missing_pct_kepler = (missing_kepler / len(kepler_df) * 100).round(2)
missing_df_kepler = pd.DataFrame({
    'Missing Count': missing_kepler[missing_kepler > 0],
    'Percentage': missing_pct_kepler[missing_kepler > 0]
}).sort_values('Percentage', ascending=False)
display(missing_df_kepler)

print("\n" + "-" * 80)
print("Basic statistics:")
display(kepler_df.describe())

KEPLER DATASET EXPLORATION

Dataset shape: 9,564 rows × 141 columns

Column names and types:
rowid                   int64
kepid                   int64
kepoi_name             object
kepler_name            object
koi_disposition        object
                       ...   
koi_dikco_mra_err     float64
koi_dikco_mdec        float64
koi_dikco_mdec_err    float64
koi_dikco_msky        float64
koi_dikco_msky_err    float64
Length: 141, dtype: object

--------------------------------------------------------------------------------
First few rows:


rowid     kepid kepoi_name   kepler_name koi_disposition koi_vet_stat  \
0      1  10797460  K00752.01  Kepler-227 b       CONFIRMED         Done   
1      2  10797460  K00752.02  Kepler-227 c       CONFIRMED         Done   
2      3  10811496  K00753.01           NaN       CANDIDATE         Done   
3      4  10848459  K00754.01           NaN  FALSE POSITIVE         Done   
4      5  10854555  K00755.01  Kepler-664 b       CONFIRMED         Done   

  koi_vet_date koi_pdisposition  koi_score  koi_fpflag_nt  koi_fpflag_ss  \
0   2018-08-16        CANDIDATE      1.000              0              0   
1   2018-08-16        CANDIDATE      0.969              0              0   
2   2018-08-16        CANDIDATE      0.000              0              0   
3   2018-08-16   FALSE POSITIVE      0.000              0              1   
4   2018-08-16        CANDIDATE      1.000              0              0   

   koi_fpflag_co  koi_fpflag_ec        koi_disp_prov  \
0              0              0  q1_q17_dr25_sup_koi   
1              0              0  q1_q17_dr25_sup_koi   
2              0              0  q1_q17_dr25_sup_koi   
3              0              0  q1_q17_dr25_sup_koi   
4              0              0  q1_q17_dr25_sup_koi   

                                        koi_comment  koi_period  \
0                                        NO_COMMENT       9.488   
1                                        NO_COMMENT      54.418   
2                                     DEEP_V_SHAPED      19.899   
3  MOD_ODDEVEN_DV---MOD_ODDEVEN_ALT---DEEP_V_SHAPED       1.737   
4                                        NO_COMMENT       2.526   

   koi_period_err1  koi_period_err2  koi_time0bk  koi_time0bk_err1  \
0            0.000           -0.000      170.539             0.002   
1            0.000           -0.000      162.514             0.004   
2            0.000           -0.000      175.850             0.001   
3            0.000           -0.000      170.308             0.000   
4            0.000           -0.000      171.596             0.001   

   koi_time0bk_err2   koi_time0  koi_time0_err1  koi_time0_err2  koi_eccen  \
0            -0.002 2455003.539           0.002          -0.002      0.000   
1            -0.004 2454995.514           0.004          -0.004      0.000   
2            -0.001 2455008.850           0.001          -0.001      0.000   
3            -0.000 2455003.308           0.000          -0.000      0.000   
4            -0.001 2455004.596           0.001          -0.001      0.000   

   koi_eccen_err1  koi_eccen_err2  koi_longp  koi_longp_err1  koi_longp_err2  \
0             NaN             NaN        NaN             NaN             NaN   
1             NaN             NaN        NaN             NaN             NaN   
2             NaN             NaN        NaN             NaN             NaN   
3             NaN             NaN        NaN             NaN             NaN   
4             NaN             NaN        NaN             NaN             NaN   

   koi_impact  koi_impact_err1  koi_impact_err2  koi_duration  \
0       0.146            0.318           -0.146         2.958   
1       0.586            0.059           -0.443         4.507   
2       0.969            5.126           -0.077         1.782   
3       1.276            0.115           -0.092         2.406   
4       0.701            0.235           -0.478         1.655   

   koi_duration_err1  koi_duration_err2  koi_ingress  koi_ingress_err1  \
0              0.082             -0.082          NaN               NaN   
1              0.116             -0.116          NaN               NaN   
2              0.034             -0.034          NaN               NaN   
3              0.005             -0.005          NaN               NaN   
4              0.042             -0.042          NaN               NaN   

   koi_ingress_err2  koi_depth  koi_depth_err1  koi_depth_err2  koi_ror  \
0               NaN    615.800          19.500         -19.500    0


--------------------------------------------------------------------------------
Missing values:


Missing Count  Percentage
koi_ingress_err2           9564     100.000
koi_longp_err2             9564     100.000
koi_teq_err1               9564     100.000
koi_incl_err2              9564     100.000
koi_incl_err1              9564     100.000
...                         ...         ...
koi_kmag                     25       0.260
koi_hmag                     25       0.260
koi_jmag                     25       0.260
koi_rmag                      9       0.090
koi_kepmag                    1       0.010

[120 rows x 2 columns]


--------------------------------------------------------------------------------
Basic statistics:


rowid        kepid  koi_score  koi_fpflag_nt  koi_fpflag_ss  \
count 9564.000     9564.000   8054.000       9564.000       9564.000   
mean  4782.500  7690628.327      0.481          0.209          0.233   
std   2761.033  2653459.081      0.477          4.767          0.423   
min      1.000   757450.000      0.000          0.000          0.000   
25%   2391.750  5556034.250      0.000          0.000          0.000   
50%   4782.500  7906892.000      0.334          0.000          0.000   
75%   7173.250  9873066.500      0.998          0.000          0.000   
max   9564.000 12935144.000      1.000        465.000          1.000   

       koi_fpflag_co  koi_fpflag_ec  koi_period  koi_period_err1  \
count       9564.000       9564.000    9564.000         9110.000   
mean           0.198          0.120      75.671            0.002   
std            0.398          0.325    1334.744            0.008   
min            0.000          0.000       0.242            0.000   
25%            0.000          0.000       2.734            0.000   
50%            0.000          0.000       9.753            0.000   
75%            0.000          0.000      40.715            0.000   
max            1.000          1.000  129995.778            0.172   

       koi_period_err2  koi_time0bk  koi_time0bk_err1  koi_time0bk_err2  \
count         9110.000     9564.000          9110.000          9110.000   
mean            -0.002      166.183             0.010            -0.010   
std              0.008       67.919             0.023             0.023   
min             -0.172      120.516             0.000            -0.569   
25%             -0.000      132.762             0.001            -0.011   
50%             -0.000      137.225             0.004            -0.004   
75%             -0.000      170.695             0.011            -0.001   
max              0.000     1472.522             0.569            -0.000   

        koi_time0  koi_time0_err1  koi_time0_err2  koi_eccen  koi_eccen_err1  \
count    9564.000        9110.000        9110.000   9201.000           0.000   
mean  2454999.183           0.010          -0.010      0.000             NaN   
std        67.919           0.023           0.023      0.000             NaN   
min   2454953.516           0.000          -0.569      0.000             NaN   
25%   2454965.762           0.001          -0.011      0.000             NaN   
50%   2454970.225           0.004          -0.004      0.000             NaN   
75%   2455003.695           0.011          -0.001      0.000             NaN   
max   2456305.522           0.569          -0.000      0.000             NaN   

       koi_eccen_err2  koi_longp  koi_longp_err1  koi_longp_err2  koi_impact  \
count           0.000      0.000           0.000           0.000    9201.000   
mean              NaN        NaN             NaN             NaN       0.735   
std               NaN        NaN             NaN             NaN       3.349   
min               NaN        NaN             NaN             NaN       0.000   
25%               NaN        NaN             NaN             NaN       0.197   
50%               NaN        NaN             NaN             NaN       0.537   
75%               NaN        NaN             NaN             NaN       0.889   
max               NaN        NaN             NaN             NaN     100.806   

       koi_impact_err1  koi_impact_err2  koi_duration  koi_duration_err1  \
count         9110.000         9110.000      9564.000           9110.000   
mean             1.960           -0.333         5.622              0.340   
std              9.422            1.250         6.472              0.670   
min              0.000          -59.320         0.052              0.000   
25%              0.040           -0.445         2.438              0.051   
50%              0.193           -0.207         3.793              0.142   
75%              0.378           -0.046         6.276              0.350   
max             85.54

### 3.2 K2 Dataset Exploration

In [15]:
# K2: Basic information
print("=" * 80)
print("K2 DATASET EXPLORATION")
print("=" * 80)

print(f"\nDataset shape: {k2_df.shape[0]:,} rows × {k2_df.shape[1]} columns\n")

print("Column names and types:")
print(k2_df.dtypes)

print("\n" + "-" * 80)
print("First few rows:")
display(k2_df.head())

print("\n" + "-" * 80)
print("Missing values:")
missing_k2 = k2_df.isnull().sum()
missing_pct_k2 = (missing_k2 / len(k2_df) * 100).round(2)
missing_df_k2 = pd.DataFrame({
    'Missing Count': missing_k2[missing_k2 > 0],
    'Percentage': missing_pct_k2[missing_k2 > 0]
}).sort_values('Percentage', ascending=False)
display(missing_df_k2)

print("\n" + "-" * 80)
print("Basic statistics:")
display(k2_df.describe())

K2 DATASET EXPLORATION

Dataset shape: 4,004 rows × 295 columns

Column names and types:
rowid             int64
pl_name          object
hostname         object
pl_letter        object
k2_name          object
                 ...   
st_nrvc         float64
st_nspec        float64
pl_nespec       float64
pl_ntranspec    float64
pl_ndispec      float64
Length: 295, dtype: object

--------------------------------------------------------------------------------
First few rows:


rowid            pl_name        hostname pl_letter  k2_name  \
0      1        BD+20 594 b       BD+20 594         b  K2-56 b   
1      2        BD+20 594 b       BD+20 594         b  K2-56 b   
2      3        BD+20 594 b       BD+20 594         b  K2-56 b   
3      4  EPIC 201111557.01  EPIC 201111557       NaN      NaN   
4      5  EPIC 201111557.01  EPIC 201111557       NaN      NaN   

    epic_hostname      epic_candname hd_name hip_name         tic_id  \
0  EPIC 210848071  EPIC 210848071.01     NaN      NaN   TIC 26123781   
1  EPIC 210848071  EPIC 210848071.01     NaN      NaN   TIC 26123781   
2  EPIC 210848071  EPIC 210848071.01     NaN      NaN   TIC 26123781   
3  EPIC 201111557  EPIC 201111557.01     NaN      NaN  TIC 176942156   
4  EPIC 201111557  EPIC 201111557.01     NaN      NaN  TIC 176942156   

                        gaia_id  default_flag disposition  \
0    Gaia DR2 58200934326315136             0   CONFIRMED   
1    Gaia DR2 58200934326315136             0   CONFIRMED   
2    Gaia DR2 58200934326315136             1   CONFIRMED   
3  Gaia DR2 3596276829630866432             1   CANDIDATE   
4  Gaia DR2 3596276829630866432             0   CANDIDATE   

             disp_refname  sy_snum  sy_pnum  sy_mnum  cb_flag discoverymethod  \
0    Espinoza et al. 2016    1.000    1.000    0.000    0.000         Transit   
1    Espinoza et al. 2016    1.000    1.000    0.000    0.000         Transit   
2    Espinoza et al. 2016    1.000    1.000    0.000    0.000         Transit   
3  Livingston et al. 2018    1.000    0.000    0.000    0.000         Transit   
4  Livingston et al. 2018    1.000    0.000    0.000    0.000         Transit   

   disc_year                                       disc_refname disc_pubdate  \
0   2016.000  <a refstr=ESPINOZA_ET_AL__2016 href=https://ui...      2016-10   
1   2016.000  <a refstr=ESPINOZA_ET_AL__2016 href=https://ui...      2016-10   
2   2016.000  <a refstr=ESPINOZA_ET_AL__2016 href=https://ui...      2016-10   
3   2018.000  <a refstr=MAYO_ET_AL__2018 href=https://ui.ads...      2018-03   
4   2018.000  <a refstr=MAYO_ET_AL__2018 href=https://ui.ads...      2018-03   

  disc_locale disc_facility           disc_telescope   disc_instrument  \
0       Space            K2  0.95 m Kepler Telescope  Kepler CCD Array   
1       Space            K2  0.95 m Kepler Telescope  Kepler CCD Array   
2       Space            K2  0.95 m Kepler Telescope  Kepler CCD Array   
3       Space            K2  0.95 m Kepler Telescope  Kepler CCD Array   
4       Space            K2  0.95 m Kepler Telescope  Kepler CCD Array   

   rv_flag  pul_flag  ptv_flag  tran_flag  ast_flag  obm_flag  micro_flag  \
0    1.000     0.000     0.000      1.000     0.000     0.000       0.000   
1    1.000     0.000     0.000      1.000     0.000     0.000       0.000   
2    1.000     0.000     0.000      1.000     0.000     0.000       0.000   
3    0.000     0.000     0.000      1.000     0.000     0.000       0.000   
4    0.000     0.000     0.000      1.000     0.000     0.000       0.000   

   etv_flag  ima_flag  dkin_flag              soltype  pl_controv_flag  \
0     0.000     0.000      0.000  Published Confirmed            0.000   
1     0.000     0.000      0.000  Published Confirmed            0.000   
2     0.000     0.000      0.000  Published Confirmed            0.000   
3     0.000     0.000      0.000  Published Candidate            0.000   
4     0.000     0.000      0.000  Published Candidate            0.000   

                                          pl_refname  pl_orbper  \
0  <a refstr=MAYO_ET_AL__2018 href=https://ui.ads...     41.689   
1  <a refstr=ESPINOZA_ET_AL__2016 href=https://ui...     41.685   
2  <a refstr=STASSUN_ET_AL__2017 href=https://ui....     41.685   
3  <a refstr=LIVINGSTON_ET_AL__2018 href=https://...      2.302   
4  <a refstr=MAYO_ET_AL__2018 href=https://ui.ads...      2.302   

   pl_orbpererr1  pl_orbpererr2  pl_orbperlim  pl_orbsmax  pl_orbsmaxerr1  \
0    


--------------------------------------------------------------------------------
Missing values:


Missing Count  Percentage
sy_kepmagerr2             4004     100.000
sy_kepmagerr1             4004     100.000
sy_icmagerr2              4004     100.000
sy_icmagerr1              4004     100.000
sy_icmag                  4004     100.000
...                        ...         ...
disc_year                   17       0.420
discoverymethod             17       0.420
rv_flag                     17       0.420
tic_id                       4       0.100
epic_hostname                2       0.050

[289 rows x 2 columns]


--------------------------------------------------------------------------------
Basic statistics:


rowid  default_flag  sy_snum  sy_pnum  sy_mnum  cb_flag  disc_year  \
count 4004.000      4004.000 3987.000 3987.000 3987.000 3987.000   3987.000   
mean  2002.500         0.451    1.051    1.127    0.000    0.000   2017.554   
std   1156.000         0.498    0.253    1.384    0.000    0.000      2.033   
min      1.000         0.000    1.000    0.000    0.000    0.000   2011.000   
25%   1001.750         0.000    1.000    0.000    0.000    0.000   2016.000   
50%   2002.500         0.000    1.000    1.000    0.000    0.000   2018.000   
75%   3003.250         1.000    1.000    2.000    0.000    0.000   2018.500   
max   4004.000         1.000    3.000    7.000    0.000    0.000   2025.000   

       rv_flag  pul_flag  ptv_flag  tran_flag  ast_flag  obm_flag  micro_flag  \
count 3987.000  3987.000  3987.000   3987.000  3987.000  3987.000    3987.000   
mean     0.241     0.000     0.000      0.994     0.000     0.000       0.001   
std      0.428     0.000     0.000      0.077     0.000     0.000       0.022   
min      0.000     0.000     0.000      0.000     0.000     0.000       0.000   
25%      0.000     0.000     0.000      1.000     0.000     0.000       0.000   
50%      0.000     0.000     0.000      1.000     0.000     0.000       0.000   
75%      0.000     0.000     0.000      1.000     0.000     0.000       0.000   
max      1.000     0.000     0.000      1.000     0.000     0.000       1.000   

       etv_flag  ima_flag  dkin_flag  pl_controv_flag  pl_orbper  \
count  3987.000  3987.000   3987.000         3987.000   3937.000   
mean      0.000     0.000      0.000            0.000     40.476   
std       0.000     0.000      0.000            0.016   1346.668   
min       0.000     0.000      0.000            0.000      0.176   
25%       0.000     0.000      0.000            0.000      2.861   
50%       0.000     0.000      0.000            0.000      6.771   
75%       0.000     0.000      0.000            0.000     14.075   
max       0.000     0.000      0.000            1.000  83830.000   

       pl_orbpererr1  pl_orbpererr2  pl_orbperlim  pl_orbsmax  pl_orbsmaxerr1  \
count       3052.000       3052.000      3937.000     812.000         805.000   
mean          14.278         -9.203        -0.001       0.105           0.006   
std          579.450        406.258         0.042       0.283           0.041   
min            0.000     -22320.000        -1.000       0.003           0.000   
25%            0.000         -0.001         0.000       0.042           0.001   
50%            0.000         -0.000         0.000       0.067           0.001   
75%            0.001         -0.000         0.000       0.105           0.003   
max        31600.000          0.000         1.000       4.500           1.000   

       pl_orbsmaxerr2  pl_orbsmaxlim  pl_rade  pl_radeerr1  pl_radeerr2  \
count         805.000        812.000 3159.000     2874.000     2874.000   
mean           -0.006         -0.004    8.445        2.871       -2.791   
std             0.041          0.061   30.083       16.089       14.727   
min            -1.000         -1.000    0.406        0.000     -367.200   
25%            -0.003          0.000    1.800        0.150       -0.670   
50%            -0.001          0.000    2.600        0.260       -0.270   
75%            -0.001          0.000    5.878        0.639       -0.140   
max             0.000          0.000 1080.000      453.800        0.000   

       pl_radelim  pl_radj  pl_radjerr1  pl_radjerr2  pl_radjlim  pl_masse  \
count    3159.000 3159.000     2874.000     2874.000    3159.000   408.000   
mean        0.000    0.754        0.257       -0.249       0.000   119.137   
std         0.018    2.687        1.447        1.309       0.018   291.920   
min         0.000    0.036        0.000      -32.760       0.000     0.297   
25%         0.000    0.161        0.013       -0.060       0.000     6.220   
50%         0.000    0.231        0.023       -0.024       0.000    12.000   
75

### 3.3 TESS Dataset Exploration

In [16]:
# TESS: Basic information
print("=" * 80)
print("TESS DATASET EXPLORATION")
print("=" * 80)

print(f"\nDataset shape: {tess_df.shape[0]:,} rows × {tess_df.shape[1]} columns\n")

print("Column names and types:")
print(tess_df.dtypes)

print("\n" + "-" * 80)
print("First few rows:")
display(tess_df.head())

print("\n" + "-" * 80)
print("Missing values:")
missing_tess = tess_df.isnull().sum()
missing_pct_tess = (missing_tess / len(tess_df) * 100).round(2)
missing_df_tess = pd.DataFrame({
    'Missing Count': missing_tess[missing_tess > 0],
    'Percentage': missing_pct_tess[missing_tess > 0]
}).sort_values('Percentage', ascending=False)
display(missing_df_tess)

print("\n" + "-" * 80)
print("Basic statistics:")
display(tess_df.describe())

TESS DATASET EXPLORATION

Dataset shape: 7,703 rows × 87 columns

Column names and types:
rowid                  int64
toi                  float64
toipfx                 int64
tid                    int64
ctoi_alias           float64
pl_pnum                int64
tfopwg_disp           object
rastr                 object
ra                   float64
raerr1               float64
raerr2               float64
decstr                object
dec                  float64
decerr1              float64
decerr2              float64
st_pmra              float64
st_pmraerr1          float64
st_pmraerr2          float64
st_pmralim           float64
st_pmrasymerr        float64
st_pmdec             float64
st_pmdecerr1         float64
st_pmdecerr2         float64
st_pmdeclim          float64
st_pmdecsymerr       float64
pl_tranmid           float64
pl_tranmiderr1       float64
pl_tranmiderr2       float64
pl_tranmidlim          int64
pl_tranmidsymerr       int64
pl_orbper            float64
pl_orbperer

rowid      toi  toipfx        tid    ctoi_alias  pl_pnum tfopwg_disp  \
0      1 1000.010    1000   50365310  50365310.010        1          FP   
1      2 1001.010    1001   88863718  88863718.010        1          PC   
2      3 1002.010    1002  124709665 124709665.010        1          FP   
3      4 1003.010    1003  106997505 106997505.010        1          FP   
4      5 1004.010    1004  238597883 238597883.010        1          FP   

          rastr      ra  raerr1  raerr2         decstr     dec  decerr1  \
0  07h29m25.85s 112.358     NaN     NaN  -12d41m45.46s -12.696      NaN   
1  08h10m19.31s 122.580     NaN     NaN  -05d30m49.87s  -5.514      NaN   
2  06h58m54.47s 104.727     NaN     NaN  -10d34m49.64s -10.580      NaN   
3  07h22m14.39s 110.560     NaN     NaN  -25d12m25.26s -25.207      NaN   
4  08h08m42.77s 122.178     NaN     NaN  -48d48m10.12s -48.803      NaN   

   decerr2  st_pmra  st_pmraerr1  st_pmraerr2  st_pmralim  st_pmrasymerr  \
0      NaN   -5.964        0.085       -0.085       0.000          1.000   
1      NaN   -4.956        0.102       -0.102       0.000          1.000   
2      NaN   -1.462        0.206       -0.206       0.000          1.000   
3      NaN   -0.939        0.041       -0.041       0.000          1.000   
4      NaN   -4.496        0.069       -0.069       0.000          1.000   

   st_pmdec  st_pmdecerr1  st_pmdecerr2  st_pmdeclim  st_pmdecsymerr  \
0    -0.076         0.072        -0.072        0.000           1.000   
1   -15.555         0.072        -0.072        0.000           1.000   
2    -2.249         0.206        -0.206        0.000           1.000   
3     1.640         0.055        -0.055        0.000           1.000   
4     9.347         0.062        -0.062        0.000           1.000   

   pl_tranmid  pl_tranmiderr1  pl_tranmiderr2  pl_tranmidlim  \
0 2459229.630           0.002          -0.002              0   
1 2459987.949           0.002          -0.002              0   
2 2459224.688           0.001          -0.001              0   
3 2458493.396           0.005          -0.005              0   
4 2459987.047           0.004          -0.004              0   

   pl_tranmidsymerr  pl_orbper  pl_orbpererr1  pl_orbpererr2  pl_orbperlim  \
0                 1      2.171          0.000         -0.000             0   
1                 1      1.932          0.000         -0.000             0   
2                 1      1.868          0.000         -0.000             0   
3                 1      2.743          0.001         -0.001             0   
4                 1      3.573          0.000         -0.000             0   

   pl_orbpersymerr  pl_trandurh  pl_trandurherr1  pl_trandurherr2  \
0                1        2.017            0.320           -0.320   
1                1        3.166            0.647           -0.647   
2                1        1.408            0.184           -0.184   
3                1        3.167            0.642           -0.642   
4                1        3.370            1.029           -1.029   

   pl_trandurhlim  pl_trandurhsymerr  pl_trandep  pl_trandeperr1  \
0               0                  1     656.886          37.778   
1               0                  1    1286.000        1186.490   
2               0                  1    1500.000           1.758   
3               0                  1     383.410           0.782   
4               0                  1     755.000        1306.550   

   pl_trandeperr2  pl_trandeplim  pl_trandepsymerr  pl_rade  pl_radeerr1  \
0         -37.778              0                 1    5.818        1.911   
1       -1186.490              0                 1   11.215        2.624   
2          -1.758              0                 1   23.753          NaN   
3          -0.782              0                 1      NaN          NaN   
4       -1306.550              0                 1   11.311        3.247   

   pl_radeerr2  pl_radelim  pl_radesymerr  pl_insol  pl_insolerr1  \
0      


--------------------------------------------------------------------------------
Missing values:


Missing Count  Percentage
raerr1                    7703     100.000
raerr2                    7703     100.000
decerr1                   7703     100.000
decerr2                   7703     100.000
pl_eqtsymerr              7703     100.000
pl_eqtlim                 7703     100.000
pl_eqterr2                7703     100.000
pl_eqterr1                7703     100.000
pl_insolsymerr            7703     100.000
pl_insollim               7703     100.000
pl_insolerr2              7703     100.000
pl_insolerr1              7703     100.000
st_loggerr2               2271      29.480
st_loggerr1               2271      29.480
st_raderr1                1963      25.480
st_raderr2                1963      25.480
pl_radeerr1               1623      21.070
pl_radeerr2               1623      21.070
st_logg                    856      11.110
st_disterr2                707       9.180
st_disterr1                707       9.180
st_rad                     507       6.580
pl_rade                    506       6.570
st_tefferr2                474       6.150
st_tefferr1                474       6.150
pl_eqt                     311       4.040
st_dist                    215       2.790
pl_insol                   176       2.280
st_teff                    161       2.090
st_pmra                    134       1.740
st_pmraerr1                134       1.740
st_pmraerr2                134       1.740
st_pmrasymerr              134       1.740
st_pmralim                 134       1.740
st_pmdecerr1               134       1.740
st_pmdec                   134       1.740
st_pmdecerr2               134       1.740
st_pmdeclim                134       1.740
st_pmdecsymerr             134       1.740
pl_orbpererr1              131       1.700
pl_orbpererr2              131       1.700
pl_orbper                  107       1.390
pl_trandurherr2             13       0.170
pl_trandurherr1             13       0.170
pl_tranmiderr1              11       0.140
pl_tranmiderr2              11       0.140
pl_trandeperr2               6       0.080
pl_trandeperr1               6       0.080


--------------------------------------------------------------------------------
Basic statistics:


rowid      toi   toipfx            tid     ctoi_alias  pl_pnum  \
count 7703.000 7703.000 7703.000       7703.000       7703.000 7703.000   
mean  3852.000 3749.039 3749.029  245522652.752  245522652.762    1.049   
std   2223.809 2152.357 2152.358  161696778.613  161696778.613    0.273   
min      1.000  101.010  101.000       2876.000       2876.010    1.000   
25%   1926.500 1861.510 1861.500  131138867.000  131138867.010    1.000   
50%   3852.000 3736.010 3736.000  249945230.000  249945230.010    1.000   
75%   5777.500 5615.510 5615.500  354091686.000  354091686.010    1.000   
max   7703.000 7509.010 7509.000 2041563029.000 2041563029.010    5.000   

            ra  raerr1  raerr2      dec  decerr1  decerr2   st_pmra  \
count 7703.000   0.000   0.000 7703.000    0.000    0.000  7569.000   
mean   179.804     NaN     NaN    1.157      NaN      NaN    -0.567   
std    103.706     NaN     NaN   47.465      NaN      NaN    77.004   
min      0.085     NaN     NaN  -89.472      NaN      NaN -1624.050   
25%     96.070     NaN     NaN  -43.355      NaN      NaN   -10.617   
50%    161.157     NaN     NaN    4.715      NaN      NaN    -1.571   
75%    283.061     NaN     NaN   43.807      NaN      NaN     8.258   
max    359.941     NaN     NaN   89.087      NaN      NaN  2074.520   

       st_pmraerr1  st_pmraerr2  st_pmralim  st_pmrasymerr  st_pmdec  \
count     7569.000     7569.000    7569.000       7569.000  7569.000   
mean         0.229       -0.229       0.000          1.000    -9.182   
std          0.632        0.632       0.000          0.000    66.875   
min          0.015       -8.000       0.000          1.000 -1230.620   
25%          0.038       -0.080       0.000          1.000   -14.695   
50%          0.051       -0.051       0.000          1.000    -3.469   
75%          0.080       -0.038       0.000          1.000     4.909   
max          8.000       -0.015       0.000          1.000  1048.840   

       st_pmdecerr1  st_pmdecerr2  st_pmdeclim  st_pmdecsymerr  pl_tranmid  \
count      7569.000      7569.000     7569.000        7569.000    7703.000   
mean          0.223        -0.223        0.000           1.000 2459552.363   
std           0.623         0.623        0.000           0.000     615.505   
min           0.016        -8.000        0.000           1.000 2457925.865   
25%           0.038        -0.071        0.000           1.000 2459196.571   
50%           0.049        -0.049        0.000           1.000 2459585.368   
75%           0.071        -0.038        0.000           1.000 2459989.900   
max           8.000        -0.016        0.000           1.000 2460863.076   

       pl_tranmiderr1  pl_tranmiderr2  pl_tranmidlim  pl_tranmidsymerr  \
count        7692.000        7692.000       7703.000          7703.000   
mean            0.004          -0.004          0.000             1.000   
std             0.044           0.044          0.000             0.000   
min             0.000          -3.305          0.000             1.000   
25%             0.001          -0.003          0.000             1.000   
50%             0.002          -0.002          0.000             1.000   
75%             0.003          -0.001          0.000             1.000   
max             3.305          -0.000          0.000             1.000   

       pl_orbper  pl_orbpererr1  pl_orbpererr2  pl_orbperlim  pl_orbpersymerr  \
count   7596.000       7572.000       7572.000      7703.000         7703.000   
mean      17.740          0.000         -0.000         0.000            1.000   
std       97.737          0.001          0.001         0.000            0.000   
min        0.152          0.000         -0.022         0.000            1.000   
25%        2.491          0.000         -0.000         0.000            1.000   
50%        4.089          0.000         -0.000         0.000            1.000   
75%        7.924          0.000         -0.000         0.000            1.000   
max     1837.890          

## 4. Compare Dataset Schemas

Analyze columns across all three datasets to identify:
- Common columns (appear in all datasets)
- Shared columns (appear in 2 datasets)
- Unique columns (appear in only 1 dataset)

In [17]:
# Get column sets
kepler_cols = set(kepler_df.columns)
k2_cols = set(k2_df.columns)
tess_cols = set(tess_df.columns)

# Find common and unique columns
common_all = kepler_cols & k2_cols & tess_cols
kepler_k2_only = (kepler_cols & k2_cols) - tess_cols
kepler_tess_only = (kepler_cols & tess_cols) - k2_cols
k2_tess_only = (k2_cols & tess_cols) - kepler_cols
kepler_unique = kepler_cols - k2_cols - tess_cols
k2_unique = k2_cols - kepler_cols - tess_cols
tess_unique = tess_cols - kepler_cols - k2_cols

print("=" * 80)
print("SCHEMA COMPARISON")
print("=" * 80)

print(f"\n📊 Total columns:")
print(f"  - Kepler: {len(kepler_cols)}")
print(f"  - K2: {len(k2_cols)}")
print(f"  - TESS: {len(tess_cols)}")

print(f"\n✅ Common to ALL three datasets ({len(common_all)} columns):")
if common_all:
    for col in sorted(common_all):
        print(f"  - {col}")
else:
    print("  (none)")

print(f"\n🔄 Shared between Kepler & K2 only ({len(kepler_k2_only)} columns):")
if kepler_k2_only:
    for col in sorted(kepler_k2_only):
        print(f"  - {col}")
else:
    print("  (none)")

print(f"\n🔄 Shared between Kepler & TESS only ({len(kepler_tess_only)} columns):")
if kepler_tess_only:
    for col in sorted(kepler_tess_only):
        print(f"  - {col}")
else:
    print("  (none)")

print(f"\n🔄 Shared between K2 & TESS only ({len(k2_tess_only)} columns):")
if k2_tess_only:
    for col in sorted(k2_tess_only):
        print(f"  - {col}")
else:
    print("  (none)")

print(f"\n🔸 Unique to Kepler ({len(kepler_unique)} columns):")
if kepler_unique:
    for col in sorted(kepler_unique):
        print(f"  - {col}")
else:
    print("  (none)")

print(f"\n🔸 Unique to K2 ({len(k2_unique)} columns):")
if k2_unique:
    for col in sorted(k2_unique):
        print(f"  - {col}")
else:
    print("  (none)")

print(f"\n🔸 Unique to TESS ({len(tess_unique)} columns):")
if tess_unique:
    for col in sorted(tess_unique):
        print(f"  - {col}")
else:
    print("  (none)")

SCHEMA COMPARISON

📊 Total columns:
  - Kepler: 141
  - K2: 295
  - TESS: 87

✅ Common to ALL three datasets (3 columns):
  - dec
  - ra
  - rowid

🔄 Shared between Kepler & K2 only (0 columns):
  (none)

🔄 Shared between Kepler & TESS only (0 columns):
  (none)

🔄 Shared between K2 & TESS only (39 columns):
  - decstr
  - pl_eqt
  - pl_eqterr1
  - pl_eqterr2
  - pl_eqtlim
  - pl_insol
  - pl_insolerr1
  - pl_insolerr2
  - pl_insollim
  - pl_orbper
  - pl_orbpererr1
  - pl_orbpererr2
  - pl_orbperlim
  - pl_rade
  - pl_radeerr1
  - pl_radeerr2
  - pl_radelim
  - pl_trandep
  - pl_trandeperr1
  - pl_trandeperr2
  - pl_trandeplim
  - pl_tranmid
  - pl_tranmiderr1
  - pl_tranmiderr2
  - pl_tranmidlim
  - rastr
  - rowupdate
  - st_logg
  - st_loggerr1
  - st_loggerr2
  - st_logglim
  - st_rad
  - st_raderr1
  - st_raderr2
  - st_radlim
  - st_teff
  - st_tefferr1
  - st_tefferr2
  - st_tefflim

🔸 Unique to Kepler (138 columns):
  - kepid
  - kepler_name
  - kepoi_name
  - koi_bin_oedp_sig

## 5. Data Quality Analysis

Examine data quality issues for each dataset:
- Duplicates
- Missing value patterns
- Data type inconsistencies
- Outliers or anomalies

In [18]:
# Check for duplicates
print("=" * 80)
print("DATA QUALITY CHECKS")
print("=" * 80)

print("\n🔍 Duplicate Rows:")
print(f"  - Kepler: {kepler_df.duplicated().sum()} duplicates")
print(f"  - K2: {k2_df.duplicated().sum()} duplicates")
print(f"  - TESS: {tess_df.duplicated().sum()} duplicates")

print("\n📊 Missing Data Summary:")
print(f"  - Kepler: {kepler_df.isnull().sum().sum():,} total missing values")
print(f"  - K2: {k2_df.isnull().sum().sum():,} total missing values")
print(f"  - TESS: {tess_df.isnull().sum().sum():,} total missing values")

print("\n💾 Memory Usage:")
print(f"  - Kepler: {kepler_df.memory_usage(deep=True).sum() / 1024**2:.2f} MB")
print(f"  - K2: {k2_df.memory_usage(deep=True).sum() / 1024**2:.2f} MB")
print(f"  - TESS: {tess_df.memory_usage(deep=True).sum() / 1024**2:.2f} MB")

DATA QUALITY CHECKS

🔍 Duplicate Rows:
  - Kepler: 0 duplicates
  - K2: 0 duplicates
  - TESS: 0 duplicates

📊 Missing Data Summary:
  - Kepler: 237,112 total missing values
  - K2: 552,676 total missing values
  - TESS: 111,013 total missing values

💾 Memory Usage:
  - Kepler: 19.73 MB
  - K2: 16.77 MB
  - TESS: 7.10 MB


## 6. Data Cleaning Strategy

Based on the exploration above, we'll implement cleaning for each dataset:

### Key Findings:
- **No duplicate rows** in any dataset ✓
- **High missing values**: Kepler (237K), K2 (552K), TESS (111K)
- **Different schemas**: Very few common columns across all three datasets
- **Different purposes**: Kepler/K2 use KOI data, while K2/TESS use confirmed planet data

### Cleaning Approach:
1. **Remove columns with 100% missing values** (uninformative)
2. **Remove columns with >95% missing values** (likely too sparse)
3. **Standardize column names** (lowercase, underscores)
4. **Identify key columns** for each dataset
5. **Handle remaining missing values** based on column importance

In [19]:
# Function to clean individual datasets
def clean_dataset(df, dataset_name, missing_threshold=95):
    """
    Clean a dataset by removing columns with too many missing values
    
    Parameters:
    - df: DataFrame to clean
    - dataset_name: Name for logging
    - missing_threshold: Percentage threshold for removing columns (default 95%)
    
    Returns:
    - Cleaned DataFrame
    """
    print(f"\n{'='*80}")
    print(f"CLEANING {dataset_name.upper()} DATASET")
    print(f"{'='*80}")
    
    original_shape = df.shape
    print(f"Original shape: {original_shape[0]:,} rows × {original_shape[1]} columns")
    
    # 1. Identify columns with excessive missing values
    missing_pct = (df.isnull().sum() / len(df) * 100)
    cols_to_drop = missing_pct[missing_pct >= missing_threshold].index.tolist()
    
    print(f"\n📊 Columns with ≥{missing_threshold}% missing values: {len(cols_to_drop)}")
    if len(cols_to_drop) > 0 and len(cols_to_drop) <= 20:
        for col in cols_to_drop:
            print(f"   - {col}: {missing_pct[col]:.1f}% missing")
    elif len(cols_to_drop) > 20:
        print(f"   (Showing first 10 of {len(cols_to_drop)})")
        for col in cols_to_drop[:10]:
            print(f"   - {col}: {missing_pct[col]:.1f}% missing")
    
    # 2. Drop columns with excessive missing values
    df_cleaned = df.drop(columns=cols_to_drop)
    print(f"\n✂️  Removed {len(cols_to_drop)} columns with excessive missing values")
    
    # 3. Report remaining missing values
    remaining_missing = df_cleaned.isnull().sum().sum()
    total_values = df_cleaned.shape[0] * df_cleaned.shape[1]
    missing_pct_total = (remaining_missing / total_values * 100)
    
    print(f"\n📉 Remaining missing values: {remaining_missing:,} ({missing_pct_total:.2f}% of total)")
    
    # 4. Show columns with most missing values (that remain)
    remaining_missing_by_col = df_cleaned.isnull().sum()
    top_missing = remaining_missing_by_col[remaining_missing_by_col > 0].sort_values(ascending=False).head(10)
    
    if len(top_missing) > 0:
        print(f"\n🔝 Top columns with missing values (after cleaning):")
        for col, count in top_missing.items():
            pct = (count / len(df_cleaned) * 100)
            print(f"   - {col}: {count:,} ({pct:.1f}%)")
    
    # 5. Final shape
    final_shape = df_cleaned.shape
    print(f"\n✅ Final shape: {final_shape[0]:,} rows × {final_shape[1]} columns")
    print(f"   Reduced from {original_shape[1]} to {final_shape[1]} columns ({original_shape[1] - final_shape[1]} removed)")
    
    return df_cleaned

# Clean each dataset
print("Starting data cleaning process...")
print("Using 95% missing value threshold for column removal\n")

kepler_clean = clean_dataset(kepler_df.copy(), "Kepler", missing_threshold=95)
k2_clean = clean_dataset(k2_df.copy(), "K2", missing_threshold=95)
tess_clean = clean_dataset(tess_df.copy(), "TESS", missing_threshold=95)

print(f"\n{'='*80}")
print("CLEANING SUMMARY")
print(f"{'='*80}")
print(f"✓ Kepler: {kepler_df.shape[1]} → {kepler_clean.shape[1]} columns ({kepler_df.shape[0]:,} rows)")
print(f"✓ K2: {k2_df.shape[1]} → {k2_clean.shape[1]} columns ({k2_df.shape[0]:,} rows)")
print(f"✓ TESS: {tess_df.shape[1]} → {tess_clean.shape[1]} columns ({tess_df.shape[0]:,} rows)")
print(f"\n🎯 All datasets cleaned and ready for further analysis or export!")

Starting data cleaning process...
Using 95% missing value threshold for column removal


CLEANING KEPLER DATASET
Original shape: 9,564 rows × 141 columns

📊 Columns with ≥95% missing values: 19
   - koi_eccen_err1: 100.0% missing
   - koi_eccen_err2: 100.0% missing
   - koi_longp: 100.0% missing
   - koi_longp_err1: 100.0% missing
   - koi_longp_err2: 100.0% missing
   - koi_ingress: 100.0% missing
   - koi_ingress_err1: 100.0% missing
   - koi_ingress_err2: 100.0% missing
   - koi_sma_err1: 100.0% missing
   - koi_sma_err2: 100.0% missing
   - koi_incl_err1: 100.0% missing
   - koi_incl_err2: 100.0% missing
   - koi_teq_err1: 100.0% missing
   - koi_teq_err2: 100.0% missing
   - koi_model_dof: 100.0% missing
   - koi_model_chisq: 100.0% missing
   - koi_sage: 100.0% missing
   - koi_sage_err1: 100.0% missing
   - koi_sage_err2: 100.0% missing

✂️  Removed 19 columns with excessive missing values

📉 Remaining missing values: 55,396 (4.75% of total)

🔝 Top columns with missing values (a

## 7. Dataset Combination Decision

### Analysis of Schema Comparison:

Based on the schema comparison in Section 4, these datasets have **very different structures**:
- **Kepler**: 122 columns (KOI-focused with `koi_` prefixes)
- **K2**: 248 columns (Planet-focused with `pl_` and `st_` prefixes)  
- **TESS**: 75 columns (Planet-focused with `pl_` and `st_` prefixes)

**Common columns across all three**: Very few (likely just metadata like `rowid`)

### Recommended Strategy: **Keep Datasets Separate**

**Rationale:**
1. **Different data models**: Kepler uses KOI (Kepler Object of Interest) data, while K2 and TESS use confirmed planet data
2. **Different missions/timeframes**: Each telescope has unique observation characteristics
3. **Minimal column overlap**: Very few shared columns means combining would create extreme sparsity
4. **Different analysis needs**: Researchers typically analyze these datasets separately by mission

### Alternative: Create Mission Identifier
If you want a unified dataset for specific analyses, we can:
1. Select only the common columns across datasets
2. Add a `mission` column to identify the source
3. Concatenate vertically

Let's keep them separate for now and export as individual cleaned files.

In [20]:
# Decision: Keep datasets separate due to different schemas
# This preserves the unique characteristics and columns of each mission

print("="*80)
print("COMBINATION DECISION")
print("="*80)

print("\n📋 Decision: Keep datasets SEPARATE")
print("\nReasons:")
print("  1. Different data models (KOI vs confirmed planets)")
print("  2. Different missions with unique observation characteristics")
print("  3. Minimal column overlap would create excessive sparsity if combined")
print("  4. Standard practice in exoplanet research is mission-specific analysis")

print("\n✓ Each cleaned dataset will be exported separately")
print("✓ This allows for mission-specific analyses while maintaining data integrity")

# Optional: Show what columns would be common if we wanted to combine later
common_cols = set(kepler_clean.columns) & set(k2_clean.columns) & set(tess_clean.columns)
print(f"\n💡 Note: Only {len(common_cols)} columns are common across all three datasets:")
if common_cols:
    for col in sorted(common_cols):
        print(f"     - {col}")
else:
    print("     (none)")
    
print("\n🔄 If needed, you can create a unified dataset later by selecting common columns")

COMBINATION DECISION

📋 Decision: Keep datasets SEPARATE

Reasons:
  1. Different data models (KOI vs confirmed planets)
  2. Different missions with unique observation characteristics
  3. Minimal column overlap would create excessive sparsity if combined
  4. Standard practice in exoplanet research is mission-specific analysis

✓ Each cleaned dataset will be exported separately
✓ This allows for mission-specific analyses while maintaining data integrity

💡 Note: Only 3 columns are common across all three datasets:
     - dec
     - ra
     - rowid

🔄 If needed, you can create a unified dataset later by selecting common columns


## 8. Export Cleaned Data

Save the cleaned datasets to the `data/clean/` folder.

In [21]:
# Create clean data directory if it doesn't exist
clean_dir = Path('../data/clean')
clean_dir.mkdir(parents=True, exist_ok=True)

print("="*80)
print("EXPORTING CLEANED DATASETS")
print("="*80)

# Export each cleaned dataset
print(f"\n📁 Export directory: {clean_dir.absolute()}\n")

# Save Kepler
kepler_path = clean_dir / 'kepler_clean.csv'
kepler_clean.to_csv(kepler_path, index=False)
print(f"✓ Kepler dataset saved: kepler_clean.csv")
print(f"  - {kepler_clean.shape[0]:,} rows × {kepler_clean.shape[1]} columns")
print(f"  - File size: {kepler_path.stat().st_size / 1024**2:.2f} MB")

# Save K2
k2_path = clean_dir / 'k2_clean.csv'
k2_clean.to_csv(k2_path, index=False)
print(f"\n✓ K2 dataset saved: k2_clean.csv")
print(f"  - {k2_clean.shape[0]:,} rows × {k2_clean.shape[1]} columns")
print(f"  - File size: {k2_path.stat().st_size / 1024**2:.2f} MB")

# Save TESS
tess_path = clean_dir / 'tess_clean.csv'
tess_clean.to_csv(tess_path, index=False)
print(f"\n✓ TESS dataset saved: tess_clean.csv")
print(f"  - {tess_clean.shape[0]:,} rows × {tess_clean.shape[1]} columns")
print(f"  - File size: {tess_path.stat().st_size / 1024**2:.2f} MB")

# Summary
print(f"\n{'='*80}")
print("EXPORT COMPLETE")
print(f"{'='*80}")
print(f"\n🎉 All cleaned datasets have been exported to: {clean_dir.absolute()}")
print(f"\n📊 Summary:")
print(f"   - Total rows exported: {kepler_clean.shape[0] + k2_clean.shape[0] + tess_clean.shape[0]:,}")
print(f"   - Kepler: {kepler_clean.shape[1]} columns (reduced from {kepler_df.shape[1]})")
print(f"   - K2: {k2_clean.shape[1]} columns (reduced from {k2_df.shape[1]})")
print(f"   - TESS: {tess_clean.shape[1]} columns (reduced from {tess_df.shape[1]})")
print(f"\n✅ Data cleaning pipeline complete!")
print(f"   Next steps: Use these cleaned datasets for analysis, visualization, or modeling")

EXPORTING CLEANED DATASETS

📁 Export directory: /Users/jorgesandoval/Documents/current/fermix/notebooks/../data/clean

✓ Kepler dataset saved: kepler_clean.csv
  - 9,564 rows × 122 columns
  - File size: 9.42 MB
✓ Kepler dataset saved: kepler_clean.csv
  - 9,564 rows × 122 columns
  - File size: 9.42 MB

✓ K2 dataset saved: k2_clean.csv
  - 4,004 rows × 248 columns
  - File size: 6.66 MB

✓ TESS dataset saved: tess_clean.csv
  - 7,703 rows × 75 columns
  - File size: 3.47 MB

EXPORT COMPLETE

🎉 All cleaned datasets have been exported to: /Users/jorgesandoval/Documents/current/fermix/notebooks/../data/clean

📊 Summary:
   - Total rows exported: 21,271
   - Kepler: 122 columns (reduced from 141)
   - K2: 248 columns (reduced from 295)
   - TESS: 75 columns (reduced from 87)

✅ Data cleaning pipeline complete!
   Next steps: Use these cleaned datasets for analysis, visualization, or modeling

✓ K2 dataset saved: k2_clean.csv
  - 4,004 rows × 248 columns
  - File size: 6.66 MB

✓ TESS data

## Summary & Results

### ✅ Completed Workflow:
1. **Setup Complete** - Libraries imported and environment configured
2. **Data Loaded** - Successfully loaded three NASA exoplanet datasets
3. **Exploration Complete** - Examined structure, types, and quality of each dataset
4. **Schema Compared** - Identified only 3 common columns (ra, dec, rowid) across all datasets
5. **Data Cleaned** - Removed columns with ≥95% missing values
6. **Decision Made** - Keep datasets separate due to different missions and data models
7. **Export Complete** - Saved cleaned datasets to `data/clean/` folder

### 📊 Final Results:

| Dataset | Original | Cleaned | Rows | File Size |
|---------|----------|---------|------|-----------|
| **Kepler** | 141 cols | 122 cols | 9,564 | 9.42 MB |
| **K2** | 295 cols | 248 cols | 4,004 | 6.66 MB |
| **TESS** | 87 cols | 75 cols | 7,703 | 3.47 MB |

### 🎯 Key Findings:
- **No duplicates** found in any dataset
- Removed **78 total columns** with excessive missing values (≥95%)
- **Different data models**: Kepler uses KOI data, K2/TESS use confirmed planet data
- **Minimal overlap**: Only 3 columns common across all datasets
- **Kept separate**: Best practice for mission-specific exoplanet research

### 📁 Output Files:
All cleaned datasets are saved in `data/clean/`:
- `kepler_clean.csv`
- `k2_clean.csv`
- `tess_clean.csv`

### 🚀 Next Steps:
- Perform mission-specific analysis on each dataset
- Create visualizations of planet properties
- Build predictive models
- Compare discovery methods across missions

---
*Data cleaning pipeline completed successfully! The cleaned datasets are ready for analysis.*